source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,544 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file azure_iot_hub_client.c
+ * @brief -------.
+ */
+
+#include "azure_iot_hub_client.h"
+#include "azure_iot_mqtt.h"
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "azure/az_iot.h"
+#include "azure/core/az_version.h"
+
+#ifndef azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC
+    #define azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC    azureiotDEFAULT_TOKEN_TIMEOUT_IN_SEC
+#endif /* azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC */
+
+#ifndef azureiothubKEEP_ALIVE_TIMEOUT_SECONDS
+    #define azureiothubKEEP_ALIVE_TIMEOUT_SECONDS    azureiotKEEP_ALIVE_TIMEOUT_SECONDS
+#endif /* azureiothubKEEP_ALIVE_TIMEOUT_SECONDS */
+
+#ifndef azureiothubSUBACK_WAIT_INTERVAL_MS
+    #define azureiothubSUBACK_WAIT_INTERVAL_MS    azureiotSUBACK_WAIT_INTERVAL_MS
+#endif /* azureiothubSUBACK_WAIT_INTERVAL_MS */
+
+#ifndef azureiothubUSER_AGENT
+    #define azureiothubUSER_AGENT    "DeviceClientType=c%2F" AZ_SDK_VERSION_STRING "%28FreeRTOS%29"
+#endif /* azureiothubUSER_AGENT */
+
+/*
+ * Topic subscribe state
+ */
+#define azureiothubTOPIC_SUBSCRIBE_STATE_NONE      ( 0x0 )
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUB       ( 0x1 )
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUBACK    ( 0x2 )
+
+/*-----------------------------------------------------------*/
+
+static void prvMQTTProcessIncomingPublish( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                           AzureIoTMQTTPublishInfo_t * pxPublishInfo )
+{
+    uint32_t ulIndex;
+
+    configASSERT( pxPublishInfo != NULL );
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    for( ulIndex = 0; ulIndex < azureiothubSUBSCRIBE_FEATURE_COUNT; ulIndex++ )
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ ulIndex ];
+
+        if( ( pxContext->_internal.pxProcessFunction != NULL ) &&
+            ( pxContext->_internal.pxProcessFunction( pxContext,
+                                                      pxAzureIoTHubClient,
+                                                      ( void * ) pxPublishInfo ) == eAzureIoTHubClientSuccess ) )
+        {
+            break;
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void prvMQTTProcessResponse( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                    AzureIoTMQTTPacketInfo_t * pxIncomingPacket,
+                                    uint16_t usPacketId )
+{
+    uint32_t ulIndex;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    if( ( pxIncomingPacket->type & 0xF0U ) == AZURE_IOT_MQTT_PACKET_TYPE_SUBACK )
+    {
```

Comment:
this is not required as line 104 does that check  .... you can add assert just to be sure.